### DISCUSSION TITLE
tRPC Standalone Middleware Changing Preceding Input Validator TypeScript Infered Type?

### DISCUSSION THREAD

**Discussion by anatolzak** [trpc/trpc#5051] [Q&A]
**Edit**
**This issue has been solved in a new release**

I have a `Project` and an `Organization` database resources, with 2 procedures: `createProject`, `updateProject`.

When creating a `Project`, you must pass a required `organizationId`. 
When updating a `Project`, you can optionally pass an `organizationId`.

I want to create a standalone middleware `organizationAuth` that is responsible for making sure the user has access to the provided `organizationId`. So I created a standalone middleware which enforces having an optional `organizationId` property in the `input` so that I can both validate a required `organizationId` when creating a project, and an optional `organizationId` when updating a project.

I would expect that in the `createProject` procedure, that if I have an input validator that requires `organizationId` and then use the `organizationAuth` middleware, then in the mutation, the `organizationId` will also be always a `string`. 

However, using the `organizationAuth` alters the type of the `organizationId` to `string | undefined`, even though we had already verified that `organizationId` will only be a string from the input validator.

Here is the sample code for everything I have described:

```typescript
const organizationAuth = standaloneMiddleware<{
  input: { organizationId?: string };
}>().create(async ({ next, ctx }) => {
  return await next({ ctx });
});

export const projectRouter = router({
  createProject: protectedProcedure
    .input(z.object({ organizationId: z.string().uuid() }))
    .use(organizationAuth)
    .mutation(({ input }) => {
      /*
        input: {
          organizationId: string | undefined;
        }
      */
      // organizationId can be undefined even though it's required in our input validator!
    }),
  updateProject: protectedProcedure
    .input(z.object({ organizationId: z.string().uuid().optional() }))
    .use(organizationAuth)
    .mutation(() => {}),
});
```

One workaround I have found is to create a separate function that given an `organizationId` and a `userId`, throws a `TRPCError` if the user doesn't have permission to access the organization. And then to create 2 separate middlewares: `organizationAuth` and `optionalOrganizationAuth`. But it seems unnecessary and repetitive since I would expect we could do all this with just 1 middleware. 

```typescript diff
const verifyOrganizationUser = async (organizationId: string, userId: string) => {
  const organizationUser = await db.organizationUser.findOne({ organizationId, userId });
  if (!organizationUser) throw new TRPCError({ code: "NOT_FOUND", message: "User is not part of this organization" });
};

const organizationAuth = standaloneMiddleware<{
  input: { organizationId: string };
}>().create(async ({ next, ctx, input: { organizationId } }) => {
  await verifyOrganizationUser({ organizationId, userId: "1" });
  return await next({ ctx });
});

const optionalOrganizationAuth = standaloneMiddleware<{
  input: { organizationId?: string };
}>().create(async ({ next, ctx, input: { organizationId } }) => {
  if (organizationId) {
    await verifyOrganizationUser({ organizationId, userId: "1" });
  }
  return await next({ ctx });
});

export const projectRouter = router({
  createProject: protectedProcedure
    .input(z.object({ organizationId: z.string().uuid() }))
    .use(organizationAuth)
    .mutation(() => {}),
  updateProject: protectedProcedure
    .input(z.object({ organizationId: z.string().uuid().optional() }))
    .use(organizationAuth)
    .use(optionalOrganizationAuth)
    .mutation(() => {}),
});
```

How would you handle this situation? Any feedback is highly appreciated! :)

**Comment by jussisaurio** [comment]
Hi. This is probably just another oversight in the implementation of the feature since it's still experimental and not very battle tested. I think the inference is just wrong here; if there's a non-optional input parser for organizationId then it should be inferred as non-optional further down the chain.

**Reply by jussisaurio** [reply]
Thanks for your examples. They will probably end up seeing a new life as part of the tRPC regression test set :)

**Reply by anatolzak** [reply]
Thank you!

**Reply by jussisaurio** [reply]
https://github.com/trpc/trpc/pull/5057

**Comment by KATT** [comment]
I have a very similar setup but I don't use the standalone middleware myself but rather something as specified here in [reusable base procedures](https://trpc.io/docs/server/procedures#reusable-base-procedures)

**Reply by anatolzak** [reply]
thanks for your feedback! How would you actually use base procedures in my example? 

If I were to create a base procedure, I would want it to fit the verification need of `organizationId` of both `createProject` and `updateProject` procedures which means that `organizationId` should be an optional field since it's required in `createProject` but optional in `updateProject`. So then I would need to revalidate the `organizationId` field as required in the `createProject`, meaning at the end `createProject` ends up validating the `organizationId` field twice, once as optional (in the base procedure) and once as required.

Is that how you would approach this situation?

```typescript
const scopedOptionalOrganization = protectedProcedure
  .input(z.object({ organizationId: z.string().uuid().optional() }))
  .use(async ({ next, ctx }) => {
    return await next({ ctx });
  });

export const projectRouter = router({
  createProject: scopedOptionalOrganization
    .input(z.object({ organizationId: z.string().uuid() }))
    .mutation(() => {}),
  updateProject: scopedOptionalOrganization.mutation(() => {}),
});
```

I greatly appreciate your feedback!

**Reply by KATT** [reply]
Ah yeah I missed the optionality detail - in my case it's never optional and the org procedure middlewares also does auth check based on the org id and the current user. It's also good to pass the org id to the db query to sure that any record created/read/updated/deleted actually matched the passed org id, treating it as a composite id
